[LCP] Implement `loadTime` for image paints

smfr opened the pull request:
#### 0c32fdb167371576029b2ffb6d99aa9f720989c8
<pre>
[LCP] Implement `loadTime` for image paints
<a href="https://bugs.webkit.org/show_bug.cgi?id=299556">https://bugs.webkit.org/show_bug.cgi?id=299556</a>
<a href="https://rdar.apple.com/161361009">rdar://161361009</a>

Reviewed by Ryosuke Niwa.

The LargestContentfulPaint entry for an image includes `loadTime`[1], which indicates the time
at which the image became available to paint, as specified by paint-timing [2]. This is not the resource load time, and there&apos;s
some ambiguity about what it means for elements that are appended to the DOM after the image
load is complete [3], but we&apos;ll track it as the time that `CachedImageClient::imageChanged()` is called, and if the
image is already available for a new renderer, use the paint time.

Bottleneck through `Document::didLoadImage()` which checks the setting.

Have RenderBox, RenderInline, RenderSVGImage and LegacyRenderSVGImage call imageDidLoad.

Move `findLayerUsedImage()` to StyleFillLayers since both RenderBox and RenderInline use it now.

Track `loadTime` and match it with a subsequent image paint by having `LargestContentfulPaintData`
store it in the per-element image-&gt;PendingImageData hash set.

[1] <a href="https://w3c.github.io/largest-contentful-paint/#dom-largestcontentfulpaint-loadtime">https://w3c.github.io/largest-contentful-paint/#dom-largestcontentfulpaint-loadtime</a>
[2] <a href="https://w3c.github.io/paint-timing/#process-image-that-finished-loading">https://w3c.github.io/paint-timing/#process-image-that-finished-loading</a>
[3] <a href="https://github.com/w3c/largest-contentful-paint/issues/146">https://github.com/w3c/largest-contentful-paint/issues/146</a>

Tested by `-platform-tests/largest-contentful-paint` tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/animated/observe-non-animated-image.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/contracted-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/cross-origin-image.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/expanded-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-TAO.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-inside-svg-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-not-fully-visible-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-removed-before-load-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-src-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/invisible-images-composited-1-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/invisible-images-composited-2-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/loadTime-after-appendChild-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/multiple-image-same-src-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/multiple-redirects-TAO-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-after-untrusted-scroll-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-css-generated-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-background-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-data-uri-background-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-data-uri-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/placeholder-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/progressively-loaded-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/redirects-tao-star-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/repeated-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/resized-image-not-reconsidered-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/same-origin-redirects-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/toJSON-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/video-poster-expected.txt: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didLoadImage const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaintData.cpp:
(WebCore::LargestContentfulPaintData::potentiallyAddLargestContentfulPaintEntry):
(WebCore::LargestContentfulPaintData::takePendingEntry):
(WebCore::LargestContentfulPaintData::didLoadImage):
(WebCore::LargestContentfulPaintData::didPaintImage):
* Source/WebCore/page/LargestContentfulPaintData.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayerImpl const): Call currentFrameIsComplete() to only count fully
renderable images for LCP.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::imageChanged):
(WebCore::findLayerUsedImage): Deleted.
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::imageChanged):
(WebCore::RenderImage::paintReplaced):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::imageChanged):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::imageChanged):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::imageChanged):
* Source/WebCore/style/Styleable.h:
(WebCore::Styleable::protectedElement const): Add to fix some Safer Cpp warnings.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
(WebCore::Style::FillLayers&lt;Layer&gt;::findLayerUsedImage const):

Canonical link: <a href="https://commits.webkit.org/300723@main">https://commits.webkit.org/300723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed267308b06f2d331ef63533718c9519f4ee6f9d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123623 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43338 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34034 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130391 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75768 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2907bd24-94c7-4d06-ad8c-f9fbb0348e7a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125500 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44061 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51932 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94002 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c17d9b68-3b8b-4d77-8429-b2bc5e5a64d2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35103 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74603 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28755 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73870 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104824 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28978 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133085 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50574 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38506 "Found 1 new test failure: imported/w3c/web-platform-tests/storage-access-api/hasStorageAccess-insecure.sub.window.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102474 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102314 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26008 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47668 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25908 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56190 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49902 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53249 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51577 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/svg/RenderSVGImage.cpp:
```diff
@@ -363,6 +363,11 @@ void RenderSVGImage::imageChanged(WrappedImagePtr newImage, const IntRect* rect)
     if (CheckedPtr cache = document().existingAXObjectCache())
         cache->deferRecomputeIsIgnoredIfNeeded(protectedImageElement().ptr());
+
+    if (auto* image = imageResource().cachedImage(); image && image->currentFrameIsComplete(this)) {
```
Can we use `RefPtr` instead?

smfr replied:
`CachedResource` is not refcounted! I'm not sure how to write safe code for it.

rniwa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
```diff
@@ -293,6 +293,11 @@ void LegacyRenderSVGImage::imageChanged(WrappedImagePtr, const IntRect*)
 
     repaint();
+
+    if (auto* image = imageResource().cachedImage(); image && image->currentFrameIsComplete(this)) {
```
Can we use `RefPtr` instead?

rniwa commented on Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
```diff
@@ -176,6 +177,34 @@ bool FillLayers<Layer>::hasAnyBackgroundClipText() const
 }
 
+template<typename Layer>
+RefPtr<StyleImage> FillLayers<Layer>::findLayerUsedImage(WrappedImagePtr image, bool& isNonEmpty) const
```
Maybe we can make this function return a pair Instead of making this function take an out argument?

smfr replied:
I did consider that, but only one caller cares about the bool, and it's a it of a hack that I'd like to clean up later.

smfr committed:
[LCP] Implement `loadTime` for image paints
https://bugs.webkit.org/show_bug.cgi?id=299556
rdar://161361009

Reviewed by Ryosuke Niwa.

The LargestContentfulPaint entry for an image includes `loadTime`[1], which indicates the time
at which the image became available to paint, as specified by paint-timing [2]. This is not the resource load time, and there's
some ambiguity about what it means for elements that are appended to the DOM after the image
load is complete [3], but we'll track it as the time that `CachedImageClient::imageChanged()` is called, and if the
image is already available for a new renderer, use the paint time.

Bottleneck through `Document::didLoadImage()` which checks the setting.

Have RenderBox, RenderInline, RenderSVGImage and LegacyRenderSVGImage call imageDidLoad.

Move `findLayerUsedImage()` to StyleFillLayers since both RenderBox and RenderInline use it now.

Track `loadTime` and match it with a subsequent image paint by having `LargestContentfulPaintData`
store it in the per-element image->PendingImageData hash set.

[1] https://w3c.github.io/largest-contentful-paint/#dom-largestcontentfulpaint-loadtime
[2] https://w3c.github.io/paint-timing/#process-image-that-finished-loading
[3] https://github.com/w3c/largest-contentful-paint/issues/146

Tested by `-platform-tests/largest-contentful-paint` tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/animated/observe-non-animated-image.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/contracted-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/cross-origin-image.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/expanded-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-TAO.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-inside-svg-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-not-fully-visible-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-removed-before-load-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-src-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/invisible-images-composited-1-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/invisible-images-composited-2-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/loadTime-after-appendChild-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/multiple-image-same-src-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/multiple-redirects-TAO-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-after-untrusted-scroll-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-css-generated-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-background-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-data-uri-background-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-data-uri-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-svg-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/placeholder-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/progressively-loaded-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/redirects-tao-star-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/repeated-image-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/resized-image-not-reconsidered-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/same-origin-redirects-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/toJSON-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/video-poster-expected.txt: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didLoadImage const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaintData.cpp:
(WebCore::LargestContentfulPaintData::potentiallyAddLargestContentfulPaintEntry):
(WebCore::LargestContentfulPaintData::takePendingEntry):
(WebCore::LargestContentfulPaintData::didLoadImage):
(WebCore::LargestContentfulPaintData::didPaintImage):
* Source/WebCore/page/LargestContentfulPaintData.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayerImpl const): Call currentFrameIsComplete() to only count fully
renderable images for LCP.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::imageChanged):
(WebCore::findLayerUsedImage): Deleted.
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::imageChanged):
(WebCore::RenderImage::paintReplaced):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::imageChanged):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::imageChanged):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::imageChanged):
* Source/WebCore/style/Styleable.h:
(WebCore::Styleable::protectedElement const): Add to fix some Safer Cpp warnings.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
(WebCore::Style::FillLayers<Layer>::findLayerUsedImage const):

Canonical link: https://commits.webkit.org/300723@main
